File: cli-node/src/core/git-analyzer.ts
import simpleGit, { SimpleGit } from 'simple-git';
import { CommitInfo, BlameEntry } from '../types.js';

export interface FileHotspot {
  file: string;
  commitCount: number;
  contributors: string[];
  lastModified: Date;
  topMessage: string;
}

/**
 * Git 数据采集器，封装 blame / log / diff-tree 等底层命令，
 * 提供面向"代码段"的历史查询能力。
 */
export class GitAnalyzer {
  private git: SimpleGit;

  constructor(repoPath: string) {
    this.git = simpleGit(repoPath);
  }

  /**
   * 获取指定行范围的 porcelain blame 信息。
   * porcelain 格式便于机器解析，每条记录包含 hash、作者、时间戳。
   */
  async getBlame(file: string, startLine: number, endLine: number): Promise<BlameEntry[]> {
    const raw = await this.git.raw([
      'blame',
      '-L', `${startLine},${endLine}`,
      '--porcelain',
      file,
    ]);
    return this.parsePorcelainBlame(raw);
  }

  /**
   * 使用 git log -L 追踪指定行范围的变更历史。
   * -L 参数让 git 自动处理行号漂移，比逐 commit diff 更准确。
   */
  async getLineHistory(file: string, startLine: number, endLine: number): Promise<CommitInfo[]> {
    const raw = await this.git.raw([
      'log',
      '--pretty=format:%H|%h|%an|%aI|%s',
      '-L', `${startLine},${endLine}:${file}`,
    ]);
    return await this.parseLogOutput(raw);
  }

  /**
   * 获取整个文件的变更历史。
   * --follow 确保文件重命名后仍能追踪到旧路径的 commit。
   */
  async getFileHistory(file: string): Promise<CommitInfo[]> {
    const raw = await this.git.raw([
      'log',
      '--pretty=format:%H|%h|%an|%aI|%s',
      '--follow',
      '--', file,
    ]);
    return await this.parseLogOutput(raw);
  }

  /**
   * 扫描目录下所有文件的修改频率，生成热点排名。
   * 解析 --name-only 输出，按文件聚合 commit 数和贡献者。
   */
  async getDirectoryHotspots(dir: string): Promise<FileHotspot[]> {
    const raw = await this.git.raw([
      'log',
      '--pretty=format:%H|%h|%an|%aI|%s',
      '--name-only',
      '--', dir,
    ]);

    const fileMap = new Map<string, {
      commits: Set<string>;
      authors: Set<string>;
      lastDate: Date;
      lastMessage: string;
    }>();

    let currentCommit: { author: string; date: Date; message: string } | null = null;

    for (const line of raw.split('\n')) {
      if (line.includes('|')) {
        const [, , author, dateStr, message] = line.split('|');
        currentCommit = { author, date: new Date(dateStr), message };
      } else if (line.trim() && currentCommit) {
        const file = line.trim();
        if (!file.startsWith(dir.replace(/\\/g, '/'))) continue;

        if (!fileMap.has(file)) {
          fileMap.set(file, {
            commits: new Set(),
            authors: new Set(),
            lastDate: currentCommit.date,
            lastMessage: currentCommit.message,
          });
        }

        const entry = fileMap.get(file)!;
        entry.commits.add(currentCommit.message);
        entry.authors.add(currentCommit.author);
        if (currentCommit.date > entry.lastDate) {
          entry.lastDate = currentCommit.date;
          entry.lastMessage = currentCommit.message;
        }
      }
    }

    const hotspots: FileHotspot[] = [];
    for (const [file, data] of fileMap) {
      hotspots.push({
        file,
        commitCount: data.commits.size,
        contributors: [...data.authors],
        lastModified: data.lastDate,
        topMessage: data.lastMessage,
      });
    }

    return hotspots.sort((a, b) => b.commitCount - a.commitCount);
  }

  async getCommitStats(hash: string): Promise<{ filesChanged: number; insertions: number; deletions: number }> {
    const raw = await this.git.raw(['diff-tree', '--shortstat', '--no-commit-id', hash]);
    const match = raw.match(/(\d+) files? changed(?:, (\d+) insertions?)?(?:, (\d+) deletions?)?/);
    if (!match) return { filesChanged: 0, insertions: 0, deletions: 0 };
    return {
      filesChanged: parseInt(match[1]) || 0,
      insertions: parseInt(match[2]) || 0,
      deletions: parseInt(match[3]) || 0,
    };
  }

  async isGitRepo(): Promise<boolean> {
    try {
      await this.git.revparse(['--is-inside-work-tree']);
      return true;
    } catch {
      return false;
    }
  }

  async getRepoRoot(): Promise<string> {
    return (await this.git.revparse(['--show-toplevel'])).trim();
  }

  async getFileLineCount(file: string): Promise<number> {
    const { readFileSync } = await import('node:fs');
    const content = readFileSync(file, 'utf-8');
    return content.split('\n').length;
  }

  // --- private helpers ---

  private async parseLogOutput(raw: string): Promise<CommitInfo[]> {
    const commits: CommitInfo[] = [];
    const seen = new Set<string>();

    for (const line of raw.split('\n')) {
      if (!line.includes('|')) continue;
      const [hash, shortHash, author, dateStr, message] = line.split('|');
      if (seen.has(hash)) continue;
      seen.add(hash);

      const stats = await this.getCommitStats(hash);
      commits.push({
        hash,
        shortHash,
        author,
        date: new Date(dateStr),
        message,
        ...stats,
      });
    }

    return commits.sort((a, b) => b.date.getTime() - a.date.getTime());
  }

  /**
   * 解析 git blame --porcelain 输出。
   * 格式: 40字符hash + 原始行号 + 最终行号，随后若干 header 行，
   * 以 TAB 开头的行为实际代码内容。
   */
  private parsePorcelainBlame(raw: string): BlameEntry[] {
    const entries: BlameEntry[] = [];
    const lines = raw.split('\n');
    let i = 0;

    while (i < lines.length) {
      const headerMatch = lines[i]?.match(/^([a-f0-9]{40}) \d+ (\d+)/);
      if (!headerMatch) { i++; continue; }

      const hash = headerMatch[1];
      const line = parseInt(headerMatch[2]);
      let author = '';
      let timestamp = 0;
      i++;

      while (i < lines.length && !lines[i].startsWith('\t')) {
        if (lines[i].startsWith('author ')) author = lines[i].slice(7);
        if (lines[i].startsWith('author-time ')) timestamp = parseInt(lines[i].slice(12));
        i++;
      }

      const content = lines[i]?.slice(1) || '';
      entries.push({ line, hash, author, date: new Date(timestamp * 1000), content });
      i++;
    }

    return entries;
  }
}

File: cli-python/src/code_archaeology/core/git_analyzer.py
from __future__ import annotations

import re
import subprocess
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Optional

from ..types import CommitInfo, BlameEntry


@dataclass
class FileHotspot:
    file: str
    commit_count: int
    contributors: list
    last_modified: Optional[datetime] = None
    top_message: str = ""


class GitAnalyzer:
    def __init__(self, repo_path: str = "."):
        self.repo_path = Path(repo_path).resolve()

    def is_git_repo(self) -> bool:
        try:
            subprocess.run(
                ["git", "rev-parse", "--is-inside-work-tree"],
                cwd=self.repo_path, capture_output=True, check=True,
            )
            return True
        except (subprocess.CalledProcessError, FileNotFoundError):
            return False

    def get_line_history(self, file: str, start_line: int, end_line: int) -> list:
        result = subprocess.run(
            ["git", "log", f"--pretty=format:%H|%h|%an|%aI|%s",
             "-L", f"{start_line},{end_line}:{file}"],
            cwd=self.repo_path, capture_output=True, text=True,
        )
        if result.returncode != 0:
            return []
        return self._parse_log_output(result.stdout)

    def get_file_history(self, file: str) -> list:
        result = subprocess.run(
            ["git", "log", "--pretty=format:%H|%h|%an|%aI|%s",
             "--follow", "--", file],
            cwd=self.repo_path, capture_output=True, text=True,
        )
        if result.returncode != 0:
            return []
        return self._parse_log_output(result.stdout)

    def get_directory_hotspots(self, dir_path: str) -> list:
        result = subprocess.run(
            ["git", "log", "--pretty=format:%H|%h|%an|%aI|%s",
             "--name-only", "--", dir_path],
            cwd=self.repo_path, capture_output=True, text=True,
        )
        if result.returncode != 0:
            return []

        file_map: dict = {}
        current_commit = None
        dir_normalized = dir_path.replace("\\", "/")

        for line in result.stdout.split("\n"):
            if "|" in line:
                parts = line.split("|", 4)
                if len(parts) >= 5:
                    current_commit = {
                        "author": parts[2],
                        "date": datetime.fromisoformat(parts[3]),
                        "message": parts[4],
                    }
            elif line.strip() and current_commit:
                file = line.strip()
                if not file.startswith(dir_normalized):
                    continue
                if file not in file_map:
                    file_map[file] = {
                        "commits": set(),
                        "authors": set(),
                        "last_date": current_commit["date"],
                        "last_message": current_commit["message"],
                    }
                entry = file_map[file]
                entry["commits"].add(current_commit["message"])
                entry["authors"].add(current_commit["author"])
                if current_commit["date"] > entry["last_date"]:
                    entry["last_date"] = current_commit["date"]
                    entry["last_message"] = current_commit["message"]

        hotspots = []
        for file, data in file_map.items():
            hotspots.append(FileHotspot(
                file=file,
                commit_count=len(data["commits"]),
                contributors=list(data["authors"]),
                last_modified=data["last_date"],
                top_message=data["last_message"],
            ))

        hotspots.sort(key=lambda h: h.commit_count, reverse=True)
        return hotspots

    def get_file_line_count(self, file: str) -> int:
        path = Path(file)
        if not path.is_absolute():
            path = self.repo_path / path
        return len(path.read_text(encoding="utf-8").split("\n"))

    def get_blame(self, file: str, start_line: int, end_line: int) -> list:
        result = subprocess.run(
            ["git", "blame", "-L", f"{start_line},{end_line}", "--porcelain", file],
            cwd=self.repo_path, capture_output=True, text=True,
        )
        if result.returncode != 0:
            return []
        return self._parse_porcelain_blame(result.stdout)

    def _parse_log_output(self, stdout: str) -> list:
        commits = []
        seen: set = set()

        for line in stdout.split("\n"):
            if "|" not in line:
                continue
            parts = line.split("|", 4)
            if len(parts) < 5:
                continue
            hash_, short_hash, author, date_str, message = parts
            if hash_ in seen:
                continue
            seen.add(hash_)

            stats = self._get_commit_stats(hash_)
            commits.append(CommitInfo(
                hash=hash_,
                short_hash=short_hash,
                author=author,
                date=datetime.fromisoformat(date_str),
                message=message,
                **stats,
            ))

        commits.sort(key=lambda c: c.date, reverse=True)
        return commits

    def _get_commit_stats(self, hash_: str) -> dict:
        result = subprocess.run(
            ["git", "diff-tree", "--shortstat", "--no-commit-id", hash_],
            cwd=self.repo_path, capture_output=True, text=True,
        )
        match = re.search(
            r"(\d+) files? changed(?:, (\d+) insertions?)?(?:, (\d+) deletions?)?",
            result.stdout,
        )
        if not match:
            return {"files_changed": 0, "insertions": 0, "deletions": 0}
        return {
            "files_changed": int(match.group(1) or 0),
            "insertions": int(match.group(2) or 0),
            "deletions": int(match.group(3) or 0),
        }

    def _parse_porcelain_blame(self, raw: str) -> list:
        entries = []
        lines = raw.split("\n")
        i = 0

        while i < len(lines):
            header = re.match(r"^([a-f0-9]{40}) \d+ (\d+)", lines[i])
            if not header:
                i += 1
                continue

            hash_ = header.group(1)
            line_num = int(header.group(2))
            author = ""
            timestamp = 0
            i += 1

            while i < len(lines) and not lines[i].startswith("\t"):
                if lines[i].startswith("author "):
                    author = lines[i][7:]
                if lines[i].startswith("author-time "):
                    timestamp = int(lines[i][12:])
                i += 1

            content = lines[i][1:] if i < len(lines) else ""
            entries.append(BlameEntry(
                line=line_num, hash=hash_, author=author,
                date=datetime.fromtimestamp(timestamp), content=content,
            ))
            i += 1

        return entries
